File: src/main/cpp/subsystems/subsystem_Elevator.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "subsystems/subsystem_Elevator.h"
#include <frc2/command/Commands.h>
#include <frc/smartdashboard/SmartDashboard.h>

subsystem_Elevator::subsystem_Elevator() : m_ElevatorMotor{ElevatorConstants::kElevatorID, rev::CANSparkMaxLowLevel::MotorType::kBrushless},
                                            m_ElevatorFollower{ElevatorConstants::kElevatorFollowerID, rev::CANSparkMaxLowLevel::MotorType::kBrushless},
                                            m_ElevatorEncoder{ElevatorConstants::kEncoderID},
                                            m_ElevatorMotorPID{m_ElevatorMotor.GetPIDController()},
                                            m_ElevatorFollowerPID{m_ElevatorFollower.GetPIDController()},
                                            m_MotorEncoder{m_ElevatorMotor.GetEncoder()},
                                            m_MotorFollowerEncoder{m_ElevatorFollower.GetEncoder()}

                                            
{
    m_ElevatorMotor.SetIdleMode(rev::CANSparkMax::IdleMode::kBrake);
    m_ElevatorFollower.SetIdleMode(rev::CANSparkMax::IdleMode::kBrake);

    m_ElevatorMotor.SetInverted(true);
    m_ElevatorFollower.SetInverted(true);
    
    m_ElevatorMotor.SetSmartCurrentLimit(5,35,117.157);
    m_ElevatorFollower.SetSmartCurrentLimit(5,35,117.157);
    //5 sec
    
    //latency or whatever
    m_ElevatorFollower.Follow(m_ElevatorMotor);
    

    // Up is Slot 0
    // Down is Slot 1

    m_ElevatorMotorPID.SetP(ElevatorConstants::kUpElevatorP, 0);
    m_ElevatorMotorPID.SetI(ElevatorConstants::kUpElevatorI, 0);
    m_ElevatorMotorPID.SetD(ElevatorConstants::kUpElevatorD, 0);

    m_ElevatorFollowerPID.SetP(ElevatorConstants::kUpElevatorP, 0);
    m_ElevatorFollowerPID.SetI(ElevatorConstants::kUpElevatorI, 0);
    m_ElevatorFollowerPID.SetD(ElevatorConstants::kUpElevatorD, 0);

    m_ElevatorMotorPID.SetP(ElevatorConstants::kDownElevatorP, 1);
    m_ElevatorMotorPID.SetI(ElevatorConstants::kDownElevatorI, 1);
    m_ElevatorMotorPID.SetD(ElevatorConstants::kDownElevatorD, 1);

    m_ElevatorFollowerPID.SetP(ElevatorConstants::kDownElevatorP, 1);
    m_ElevatorFollowerPID.SetI(ElevatorConstants::kDownElevatorI, 1);
    m_ElevatorFollowerPID.SetD(ElevatorConstants::kDownElevatorD, 1);

    SetElevatorPosition(ElevatorConstants::kElevatorStow);
    frc::SmartDashboard::SmartDashboard::PutNumber("motor current", m_ElevatorMotor.GetOutputCurrent());
    frc::SmartDashboard::SmartDashboard::PutNumber("follower motor current", m_ElevatorFollower.GetOutputCurrent());
    //m_ElevatorEncoder.SetDistancePerRotation(5.5);
}

// // Will merge the command ptrs in Cgroup later w saliks wrist
// frc2::CommandPtr subsystem_Elevator::ToMidCone()
// {
//     MoveElevator(ElevatorConstants::kElevatorScoreMidCone);
//     //its givin error cuzzzz i didnt return anything to ptr
//     //return frc2::cmd::Sequence(MoveElevator(ElevatorConstants::kElevatorScoreMidCone));
//     // return frc2::cmd::Sequence()
// }

// frc2::CommandPtr subsystem_Elevator::ToHighCone()
// {
//     MoveElevator(ElevatorConstants::kElevatorScoreHighCone);
// }   

// frc2::CommandPtr subsystem_Elevator::ToMidCube()
// {
//     MoveElevator(ElevatorConstants::kElevatorScoreMidCube);
// }

// frc2::CommandPtr subsystem_Elevator::ToHighCube()
// {
//     MoveElevator(ElevatorConstants::kElevatorScoreHighCube);
// }

// frc2::CommandPtr subsystem_Elevator::ToStow()
// {
//     MoveElevator(ElevatorConstants::kElevatorStow);
// }

// frc2::CommandPtr subsystem_Elevator::ToHybrid()
// {
//     MoveElevator(ElevatorConstants::kElevatorScoreHybrid);
// }

// frc2::CommandPtr subsystem_Elevator::ToDoubleSubstation()
// {
//     MoveElevator(ElevatorConstants::kElevatorIntakeDoubleSubstation);
// }

//  : m_ElevatorEncoder(frc::DutyCycleEncoder::DutyCycleEncoder(int channel))
// void subsystem_Elevator::ElevatorEncoder()
// {
//     //initializing elevator encoder

// }

// void subsystem_Elevator::MoveElevator(double angle)
// {
    
//     //does follower need to be updated here if it .follows
//     //move set reference to periodic later if this doesn't work
//     // also change to rotations if it dont workout
//     //not using motionmagic cuz inertia of weight of gamepiece messes up so use sid
//     m_ElevatorMotorPID.SetReference(angle, rev::CANSparkMax::ControlType::kPosition, 0);
//     m_ElevatorFollowerPID.SetReference(angle, rev::CANSparkMax::ControlType::kPosition, 0);
    
// }

void subsystem_Elevator::ResetElevatorEncoder()
{
    m_ElevatorEncoder.Reset();
    //frc::SmartDashboard::SmartDashboard::PutBoolean("Reset Elevator Encoder", true);

}

void subsystem_Elevator::SetElevatorPosition(double ElevatorPositionPassedIn)
{
    DesiredElevatorPosition = ElevatorPositionPassedIn;

    if(ElevatorPositionPassedIn == ElevatorConstants::kElevatorStow)
    {
        DesiredElevatorSlot = 1;
        m_ElevatorMotorPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, ElevatorConstants::kDownElevatorSlot);
    }
    else
    {
        DesiredElevatorSlot = 0;
        m_ElevatorMotorPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, ElevatorConstants::kUpElevatorSlot);
    }

    //m_ElevatorFollowerPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, 0);
    frc::SmartDashboard::SmartDashboard::PutNumber("Elevator Position Passed in", ElevatorPositionPassedIn);
    //m_ElevatorFollowerPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, 0);
}

void subsystem_Elevator::MoveElevatorManually(double axis)
{
    // manualX = armX + (leftAxis * ArmConstants::JoystickToArm);
    // manualY = armY + (rightAxis * ArmConstants::JoystickToArm);
    // MoveArm(manualX, manualY);

    DesiredElevatorPosition += axis * ElevatorConstants::kManualSpeed;

    if(axis < 0)
    {
        DesiredElevatorSlot = 1;
        m_ElevatorMotorPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, ElevatorConstants::kDownElevatorSlot);
    }
    if(axis > 0)
    {
        DesiredElevatorSlot = 0;
        m_ElevatorMotorPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, ElevatorConstants::kUpElevatorSlot);
    }

    
    //m_ElevatorFollowerPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, 0);
}

double subsystem_Elevator::GetElevatorPosition() 
{
    return m_ElevatorEncoder.GetDistance();
}

double subsystem_Elevator::GetElevatorsMotorPosition()
{
    return m_MotorEncoder.GetPosition();
}

void subsystem_Elevator::SetElevatorMotorSpeed(double speed)
{
    m_ElevatorMotor.Set(speed);
    m_ElevatorFollower.Set(speed);
}

// this part is diff from wrist???
bool subsystem_Elevator::IsElevatorAtDesiredPosition()
{
    // if(fabs(DesiredElevatorPosition - ElevatorEnc) < ElevatorConstants::kBufferZone)
    // {
    //     return true;
    // }
    // return false;

    if(fabs(DesiredElevatorPosition - ElevatorEnc) < ElevatorConstants::kBufferZone)
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool subsystem_Elevator::ElevatorThreshold(double Threshold)
{
    if(DesiredElevatorPosition >= ElevatorEnc)
    {
        return ElevatorEnc > Threshold;
    }
    else
    {
        return ElevatorEnc < Threshold;
    }

    //return DesiredElevatorPosition >= 0.0 ? ElevatorEnc > Threshold : ElevatorEnc < Threshold;
}

// frc2::CommandPtr SetElevatorMotorSpeedCommand(std::function<double()> speed){
//     return RunOnce[this]{return SetElevatorMotorSpeed(speed());};
// }

// This method will be called once per scheduler run
void subsystem_Elevator::Periodic() {
    //m_motor2.set(m_motor1.get());
    ElevatorEnc = m_MotorEncoder.GetPosition();
    frc::SmartDashboard::SmartDashboard::PutNumber("Elevator Pos", ElevatorEnc);
    if (!IsElevatorAtDesiredPosition())
    {
        frc::SmartDashboard::SmartDashboard::PutNumber("Desired Elevator Position", DesiredElevatorPosition);
        m_ElevatorMotorPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, DesiredElevatorSlot);
        //m_ElevatorFollowerPID.SetReference(DesiredElevatorPosition, rev::CANSparkMax::ControlType::kPosition, 0);
        frc::SmartDashboard::SmartDashboard::PutBoolean("Went into if statement", true);
    }
    else
    {
        frc::SmartDashboard::SmartDashboard::PutBoolean("Went into if statement", false);
    }

    frc::SmartDashboard::SmartDashboard::PutNumber("Current motor Encoder Value", ElevatorEnc);


}

File: src/main/java/frc/robot/subsystems/subsystem_Elevator.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package frc.robot.subsystems;

import edu.wpi.first.wpilibj2.command.SubsystemBase;

import com.revrobotics.CANSparkMax;
import com.revrobotics.SparkMaxPIDController;
import com.revrobotics.CANSparkMax.ControlType;
import com.revrobotics.CANSparkMax.IdleMode;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.CANSparkMaxLowLevel.MotorType;
import edu.wpi.first.wpilibj.DutyCycleEncoder;

import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import frc.robot.Constants;
import frc.robot.Constants.ElevatorConstants;

public class subsystem_Elevator extends SubsystemBase {
  /** Creates a new subsystem_Elevator. */
  private CANSparkMax m_ElevatorMotor;
  private CANSparkMax m_ElevatorFollower;
  private DutyCycleEncoder m_ElevatorEncoder;
  private SparkMaxPIDController m_ElevatorMotorPID;
  private SparkMaxPIDController m_ElevatorFollowerPID;
  private RelativeEncoder m_ElevatorMotorEncoder;
  private RelativeEncoder m_ElevatorFollowerEncoder;

  private double desiredElevatorPosition = 0.0;
  private double elevatorEnc = 0.0;
  private int desiredElevatorSlot = 0;

  //temp for pid tuning
  
  private double kUpP = SmartDashboard.getNumber("Up P", Constants.ElevatorConstants.kUpElevatorP);
  private double kUpI = SmartDashboard.getNumber("Up I", Constants.ElevatorConstants.kUpElevatorI);
  private double kUpD = SmartDashboard.getNumber("Up D", Constants.ElevatorConstants.kUpElevatorD);

  public subsystem_Elevator() {
    m_ElevatorMotor = new CANSparkMax(Constants.ElevatorConstants.kElevatorID, MotorType.kBrushless);
    m_ElevatorFollower = new CANSparkMax(Constants.ElevatorConstants.kElevatorFollowerID, MotorType.kBrushless);
    m_ElevatorEncoder = new DutyCycleEncoder(Constants.ElevatorConstants.kEncoderID);
    m_ElevatorMotorPID = m_ElevatorMotor.getPIDController();
    m_ElevatorFollowerPID = m_ElevatorFollower.getPIDController();
    m_ElevatorMotorEncoder = m_ElevatorMotor.getEncoder();
    m_ElevatorFollowerEncoder = m_ElevatorFollower.getEncoder();
    
    m_ElevatorMotor.setIdleMode(IdleMode.kBrake);
    m_ElevatorFollower.setIdleMode(IdleMode.kBrake);
    
    m_ElevatorMotor.setIdleMode(IdleMode.kBrake);
    m_ElevatorFollower.setIdleMode(IdleMode.kBrake);

    m_ElevatorMotor.setInverted(true);
    m_ElevatorFollower.setInverted(true);

    m_ElevatorMotor.setSmartCurrentLimit(40);
    m_ElevatorFollower.setSmartCurrentLimit(40);

    m_ElevatorFollower.follow(m_ElevatorMotor);
    
    SmartDashboard.putNumber("Up P", Constants.ElevatorConstants.kUpElevatorP);
    SmartDashboard.putNumber("Up I", Constants.ElevatorConstants.kUpElevatorI);
    SmartDashboard.putNumber("Up D", Constants.ElevatorConstants.kUpElevatorD);
    
    kUpP = SmartDashboard.getNumber("Up P", Constants.ElevatorConstants.kUpElevatorP);
    kUpI = SmartDashboard.getNumber("Up I", Constants.ElevatorConstants.kUpElevatorI);
    kUpD = SmartDashboard.getNumber("Up D", Constants.ElevatorConstants.kUpElevatorD);

    m_ElevatorMotorPID.setP(Constants.ElevatorConstants.kUpElevatorP, Constants.ElevatorConstants.kUpElevatorSlot);
    m_ElevatorMotorPID.setI(Constants.ElevatorConstants.kUpElevatorI, Constants.ElevatorConstants.kUpElevatorSlot);
    m_ElevatorMotorPID.setD(Constants.ElevatorConstants.kUpElevatorD, Constants.ElevatorConstants.kUpElevatorSlot);
    
    m_ElevatorFollowerPID.setP(Constants.ElevatorConstants.kUpElevatorP, Constants.ElevatorConstants.kUpElevatorSlot);
    m_ElevatorFollowerPID.setI(Constants.ElevatorConstants.kUpElevatorI, Constants.ElevatorConstants.kUpElevatorSlot);
    m_ElevatorFollowerPID.setD(Constants.ElevatorConstants.kUpElevatorD, Constants.ElevatorConstants.kUpElevatorSlot);

    m_ElevatorMotorPID.setP(Constants.ElevatorConstants.kDownElevatorP, Constants.ElevatorConstants.kDownElevatorSlot);
    m_ElevatorMotorPID.setI(Constants.ElevatorConstants.kDownElevatorI, Constants.ElevatorConstants.kDownElevatorSlot);
    m_ElevatorMotorPID.setD(Constants.ElevatorConstants.kDownElevatorD, Constants.ElevatorConstants.kDownElevatorSlot);

    m_ElevatorFollowerPID.setP(Constants.ElevatorConstants.kDownElevatorP, Constants.ElevatorConstants.kDownElevatorSlot);
    m_ElevatorFollowerPID.setI(Constants.ElevatorConstants.kDownElevatorI, Constants.ElevatorConstants.kDownElevatorSlot);
    m_ElevatorFollowerPID.setD(Constants.ElevatorConstants.kDownElevatorD, Constants.ElevatorConstants.kDownElevatorSlot);

    setElevatorPosition(Constants.ElevatorConstants.kElevatorStow);
    SmartDashboard.putNumber("motor current", m_ElevatorMotor.getOutputCurrent());
    SmartDashboard.putNumber("follower motor current", m_ElevatorFollower.getOutputCurrent());
  }

  public void resetElevatorEncoder(){
    m_ElevatorEncoder.reset();
  }

  public void setElevatorPosition(double elevatorPosition){
    
    desiredElevatorPosition = elevatorPosition;

    if(elevatorPosition == Constants.ElevatorConstants.kElevatorStow){
      desiredElevatorSlot = Constants.ElevatorConstants.kDownElevatorSlot;
      m_ElevatorMotorPID.setReference(desiredElevatorPosition, ControlType.kPosition, desiredElevatorSlot);
    }
    else {
      desiredElevatorSlot = Constants.ElevatorConstants.kUpElevatorSlot;
      m_ElevatorMotorPID.setReference(desiredElevatorPosition, ControlType.kPosition, desiredElevatorSlot, ElevatorConstants.arbFF);
    }     

    SmartDashboard.putNumber("Elevator Position Passed In", elevatorPosition);
  }

  // public void moveElevatorManually(double axis){
    
  //   desiredElevatorPosition += axis * Constants.ElevatorConstants.kManualSpeed;
    
  //   if(axis < 0){
  //     desiredElevatorSlot = Constants.ElevatorConstants.kDownElevatorSlot;
  //     m_ElevatorMotorPID.setReference(desiredElevatorPosition, ControlType.kPosition, desiredElevatorSlot);
  //   }
  //   if(axis > 0){
  //     desiredElevatorSlot = Constants.ElevatorConstants.kUpElevatorSlot;
  //     m_ElevatorMotorPID.setReference(desiredElevatorPosition, ControlType.kPosition, desiredElevatorSlot);
  //   }
  // }

  public double getElevatorPosition(){
    return m_ElevatorEncoder.getDistance();
  }

  public double getElevatorMotorPosition(){
    return m_ElevatorMotorEncoder.getPosition();
  }

  public void setElevatorMotorSpeed(double speed){
    m_ElevatorMotor.set(speed);
    m_ElevatorFollower.set(speed);
  }

  public boolean isElevatorAtDesiredPosition(){
    return Math.abs(desiredElevatorPosition - elevatorEnc) < Constants.ElevatorConstants.kBufferZone;
  }

  public boolean elevatorThreshold(double threshold){
    return desiredElevatorPosition >= 0.0 ? elevatorEnc > threshold : elevatorEnc < threshold;
  }

  @Override
  public void periodic() {
    //temp code
    kUpP = SmartDashboard.getNumber("Up P", Constants.ElevatorConstants.kUpElevatorP);
    kUpI = SmartDashboard.getNumber("Up I", Constants.ElevatorConstants.kUpElevatorI);
    kUpD = SmartDashboard.getNumber("Up D", Constants.ElevatorConstants.kUpElevatorD);

    SmartDashboard.putNumber("Gotten Number", kUpP);

    m_ElevatorMotorPID.setP(kUpP, Constants.ElevatorConstants.kUpElevatorSlot);
    m_ElevatorMotorPID.setI(kUpI, Constants.ElevatorConstants.kUpElevatorSlot);
    m_ElevatorMotorPID.setD(kUpD, Constants.ElevatorConstants.kUpElevatorSlot);
    
    m_ElevatorFollowerPID.setP(Constants.ElevatorConstants.kUpElevatorP, Constants.ElevatorConstants.kUpElevatorSlot);
    m_ElevatorFollowerPID.setI(Constants.ElevatorConstants.kUpElevatorI, Constants.ElevatorConstants.kUpElevatorSlot);
    m_ElevatorFollowerPID.setD(Constants.ElevatorConstants.kUpElevatorD, Constants.ElevatorConstants.kUpElevatorSlot);

    // This method will be called once per scheduler run
    elevatorEnc = m_ElevatorMotorEncoder.getPosition();
    SmartDashboard.putNumber("Elevator Pos", elevatorEnc);

    if(!isElevatorAtDesiredPosition()){
      SmartDashboard.putNumber("Desired Positin", desiredElevatorPosition);
      if (desiredElevatorSlot == Constants.ElevatorConstants.kDownElevatorSlot){
        m_ElevatorMotorPID.setReference(desiredElevatorPosition, ControlType.kPosition, desiredElevatorSlot);
      } else {
        m_ElevatorMotorPID.setReference(desiredElevatorPosition, ControlType.kPosition, desiredElevatorSlot, ElevatorConstants.arbFF);
      }
      SmartDashboard.putBoolean("Went Into If-Statement", true);
    }
    else {
      SmartDashboard.putBoolean("Went Into If-Statement", false);
    }

    SmartDashboard.putNumber("Current Motor Encoder Value", elevatorEnc);
  }
}
